File: test/Concurrency.js
const child = require("child_process")
const exec = child.exec

const run = () => {
	return new Promise((resolve) => {
		exec(`sleep 1`, () => {
			return resolve()
		})
	})
}

const runAllatOnce = async (times) => {
	return new Promise((resolve) => {
		Res = []
		for (let c = 1; c <= times; c++) {
			Res.push(run())
		}
		Promise.all(Res).then(() => resolve())
	})
}

const runAll = async (times) => {
	for (let c = 1; c <= times; c++) {
		await run()
	}
}

const main = async () => {
	let startat
	console.log("started")
	startat = Date.now()
	await runAllatOnce(3000)
	console.log("done all at once in", Date.now() - startat, "ms")
	startat = Date.now()
	await runAll(1)
	console.log("done all in", Date.now() - startat, "ms")
}

main()

File: test/Concurrency.go
package main

import (
	"fmt"
	"os/exec"
	"sync"
	"time"
)

func run() {
	cmd := exec.Command("sleep", "1")
	err := cmd.Run()
	if err != nil {
		panic(err)
	}
}

func runAll(times int) {
	for i := 0; i < times; i++ {
		run()
	}
}

func runAllatOnce(times int) {
	var wg sync.WaitGroup
	for i := 0; i < times; i++ {
		wg.Add(1)
		go func() {
			defer wg.Done()
			run()
		}()
	}
	wg.Wait()
}

func main() {
	fmt.Println("started")

	var start_at time.Time

	start_at = time.Now()
	runAllatOnce(2999)
	fmt.Println("finished all at once in ", time.Now().Sub(start_at))

	start_at = time.Now()
	runAll(5)
	fmt.Println("finished all in ", time.Now().Sub(start_at))
}
